### ISSUE TITLE
[2.5] Update!! multipart/form data parsing takes 5 seconds when there are simple data parts

### ISSUE THREAD

**Issue by schmitch** [playframework/playframework#5751]
Actually when trying to upload a file via multipart/form-data while being on a Play! 2.5-RC or directly on playframework - master it takes a huge amount of time until I can start processing the multipart form-data. It looks like the new streaming body parser of multipart/form-data has a bug.

### Play Version (2.5.x / etc)

Play! 2.5-RC1, 2.5-RC2, master

### API (Scala / Java / Neither / Both)

- Both (Tested both under Mac, and Scala on Mac and Fedora)
- Akka-Http
- Netty

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)

I'm on Mac, however I had the same problem on a Linux system which I tested with 2.5-RC1

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)

Oracle JDK from 1.8.0_66 to 1.8.0_74

### Library Dependencies

I'm just using:
  "com.typesafe.play" %% "play" % playVersion,
  "com.typesafe.play" %% "play-netty-server" % playVersion,
  "com.typesafe.play" %% "play-logback" % playVersion,

or a new activator play-scala-seed with a single route.

### Expected Behavior

Please describe the expected behavior of the issue, starting from the first action.

1. upload a file via ``curl -v -F key1=value1 -F upload=@demo_file http://localhost:9003``
2. should be instantly especially if file is just 5kb (at least not 5000ms)

### Actual Behavior

1. upload a file via ``curl -v -F key1=value1 -F upload=@demo_file http://localhost:9003``
2. takes 5000ms to upload even if just doing:

```scala
  def multipart = Action { implicit r =>
    r.body.asMultipartFormData.map { mp =>

      Ok
    }.getOrElse {
      BadRequest
    }
  }
```
And only having a LoggingFilter:

```scala
class LoggingFilter(implicit val mat: Materializer) extends Filter {

  val logger = Logger(this.getClass)

  def apply(nextFilter: RequestHeader => Future[Result])
           (requestHeader: RequestHeader): Future[Result] = {

    val startTime = System.currentTimeMillis

    nextFilter(requestHeader).map { result =>

      val endTime = System.currentTimeMillis
      val requestTime = endTime - startTime


      logger.info(s"${requestHeader.method} ${requestHeader.uri} " +
        s"took ${requestTime}ms and returned ${result.header.status}")


      result.withHeaders("Request-Time" -> requestTime.toString)
    }
  }

}
```
Without the filter it takes nearly the same time.

There is another way of testing it while being in my multipart branch and adding a test like that:

https://github.com/playframework/playframework/pull/5698
https://gist.github.com/schmitch/e296dc493bd3c42407f4

Actually this will throw a timeout error if you Await.result(ws.url..., 1.seconds)
However the test finishes in 5 seconds:
```
[info] MultipartUploadSpec
[info] 
[info] MultipartFormDataUpload should
[error]   ! upload in fast
[error]    java.util.concurrent.TimeoutException: Futures timed out after [1 second] (MultipartUploadSpec.scala:48)
[error] play.api.libs.ws.multipart.MultipartUploadSpec$$anonfun$2$$anonfun$apply$3$$anonfun$apply$4.apply(MultipartUploadSpec.scala:48)
[error] play.api.libs.ws.multipart.MultipartUploadSpec$$anonfun$2$$anonfun$apply$3$$anonfun$apply$4.apply(MultipartUploadSpec.scala:45)
[error] play.api.libs.ws.multipart.MultipartUploadSpec$$anonfun$withServer$1.apply(MultipartUploadSpec.scala:41)
[error] play.api.test.PlayRunners$class.running(Helpers.scala:76)
[error] play.api.libs.ws.multipart.MultipartUploadSpec.running(MultipartUploadSpec.scala:20)
[error] play.api.libs.ws.multipart.MultipartUploadSpec.withServer(MultipartUploadSpec.scala:41)
[error] play.api.libs.ws.multipart.MultipartUploadSpec.withSimpleServer(MultipartUploadSpec.scala:35)
[error] play.api.libs.ws.multipart.MultipartUploadSpec$$anonfun$2$$anonfun$apply$3.apply(MultipartUploadSpec.scala:45)
[error] play.api.libs.ws.multipart.MultipartUploadSpec$$anonfun$2$$anonfun$apply$3.apply(MultipartUploadSpec.scala:45)
[info] 
[info] 
[info] 
[info] Total for specification MultipartUploadSpec
[info] Finished in 5 seconds, 456 ms
[info] 1 example, 2 expectations, 0 failure, 1 error
```

Also this is definitly a release blocker and should be fixed before Play! 2.5 will be released

This Bug actually **only** appears to happen when uploading a File **and** form/data:

```
schmitch@SHANGHAI:~$ curl -v -F upload=@golang.md http://localhost:9090/
*   Trying ::1...
* Connected to localhost (::1) port 9090 (#0)
> POST / HTTP/1.1
> Host: localhost:9090
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Length: 1226
> Expect: 100-continue
> Content-Type: multipart/form-data; boundary=------------------------de08532d19b8790a
> 
< HTTP/1.1 100 Continue
< Server: akka-http/2.4.2
< Date: Tue, 01 Mar 2016 21:09:03 GMT
< HTTP/1.1 200 OK
< Request-Time: 11
< Server: akka-http/2.4.2
< Date: Tue, 01 Mar 2016 21:09:03 GMT
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
```

```
schmitch@SHANGHAI:~$ curl -v -F data=value -F upload=@golang.md http://localhost:9090/
*   Trying ::1...
* Connected to localhost (::1) port 9090 (#0)
> POST / HTTP/1.1
> Host: localhost:9090
> User-Agent: curl/7.43.0
> Accept: */*
> Content-Length: 1324
> Expect: 100-continue
> Content-Type: multipart/form-data; boundary=------------------------f2fe80cb9aab7591
> 
< HTTP/1.1 100 Continue
< Server: akka-http/2.4.2
< Date: Tue, 01 Mar 2016 21:08:45 GMT
< HTTP/1.1 200 OK
< Request-Time: 5028
< Server: akka-http/2.4.2
< Date: Tue, 01 Mar 2016 21:08:50 GMT
< Content-Length: 0
< 
* Connection #0 to host localhost left intact
```

I always wondered why @jroper didn't had the problem on the Java App, but my Linux Server and my Mac had it. However on my Test App I always had a CSRF Token sent, too.

**Comment by mkurz** [comment]
I am assigning the 2.5.0 milestone - this should be checked before the release probably.

**Comment by schmitch** [comment]
nobody cares

**Comment by gaydenko** [comment]
@schmitch , hi! Have you got a github project ready for `git clone`, `sbt run` exposing the bug?

**Comment by schmitch** [comment]
@gaydenko https://github.com/schmitch/play-bug
So multipart/form data is broken in 2.5.0 great! good that I only use it in a personal application.

**Comment by gaydenko** [comment]
@schmitch , thanks! - I'll try to dig in. Just started to familiar with 2.5 ([1]), while streaming upload is very important at my case.

[1] https://groups.google.com/forum/#!folder/Current/play-framework/gXXClWrWel8

**Comment by wsargent** [comment]
FYI, I am removing all the milestones on the open bugs as they are now 2.5.x rather than "2.5.0".  

@schmitch it okay -- there will be a patch release out very shortly, as people dig into 2.5.0 more.

**Comment by schmitch** [comment]
@gaydenko that doc is prolly outdated. since it's a  
```type FilePartHandler[A] = FileInfo => Accumulator[ByteString, FilePart[A]]```

An Accumulator is a Source converted to a Future. And I would've resolved this bug, but I couldn't find it. Not sure where it lies. Since it's only happening on the DataPart and not on the FilePart.

**Comment by gaydenko** [comment]
@schmitch , thanks for the tip! - now I need some time to learn all these new (2.5) things...

I have reproduced the issue with `Run completed in 4 seconds, 80 milliseconds`.

**Comment by nlhkh** [comment]
I am seeing this same issue.

**Comment by smaass** [comment]
I have this exact same problem with:

`def foo = Action { implicit r =>
    dataForm.bindFromRequest.fold(
        formWithErrors => BadRequest,
        
        data => Ok
    )
  }`

A POST request to this endpoint completes in around 5000ms.

**Comment by marcosinigaglia** [comment]
Same issue with Java

**Comment by Enalmada** [comment]
After upgrading to 2.5.0 my multipart forms hang for a few minutes and then return "ERROR: 400 {"error":"InputMissing","message":"File is empty"}".  Any workaround?  If not, I think this release should be pulled until the fix so that anyone with consumer facing multipart forms (profile image, etc) don't get completely screwed.

**Comment by jroper** [comment]
Fixed in #5860.
